onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -194,30 +193,38 @@ void AddNonTensor<TensorSeq>(OrtValue& val, std::vector<py::object>& pyobjs) {
 
 void AddNonTensorAsPyObj(OrtValue& val, std::vector<py::object>& pyobjs) {
   // Should be in sync with core/framework/datatypes.h
-  if (val.Type() == DataTypeImpl::GetType<TensorSeq>()) {
+  auto val_type = val.Type();
+  if (val_type->IsTensorSequenceType()) {
     AddNonTensor<TensorSeq>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapStringToString>()) {
-    AddNonTensor<MapStringToString>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapStringToInt64>()) {
-    AddNonTensor<MapStringToInt64>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapStringToFloat>()) {
-    AddNonTensor<MapStringToFloat>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapStringToDouble>()) {
-    AddNonTensor<MapStringToDouble>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapInt64ToString>()) {
-    AddNonTensor<MapInt64ToString>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapInt64ToInt64>()) {
-    AddNonTensor<MapInt64ToInt64>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapInt64ToFloat>()) {
-    AddNonTensor<MapInt64ToFloat>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<MapInt64ToDouble>()) {
-    AddNonTensor<MapInt64ToDouble>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<VectorMapStringToFloat>()) {
-    AddNonTensor<VectorMapStringToFloat>(val, pyobjs);
-  } else if (val.Type() == DataTypeImpl::GetType<VectorMapInt64ToFloat>()) {
-    AddNonTensor<VectorMapInt64ToFloat>(val, pyobjs);
   } else {
-    throw std::runtime_error("Output is a non-tensor type which is not supported.");
+    utils::ContainerChecker c_checker(val_type);
```

Comment:
The new code doesn't look much different from the previous one in terms of structure - you still need all the if/elses. One major difference is the calls to GetType. I guess that's what you're trying to avoid?

Reply:
a) avoid repeated GetType<>() calls, have a flexiblity to test for any container type wo being typedefed previously or write any other code that is not dependent on what was being typedefed.

---
In reply to: [353958679](https://github.com/microsoft/onnxruntime/pull/2522#discussion_r353958679) [](ancestors = 353958679)